File: scraper/python/GrabzIt/ScrapeResult.py
#!/usr/bin/python

import cgi
import cgitb
import json
import os
import xml.etree.ElementTree as ET

class ScrapeResult:

    def __init__(self, path = None):
        cgitb.enable()
        
        self.data = None
        self.extension = None
        self.filename = None
        self.file = None
        
        if path != None:
            fo = open(path)
            self.data = fo.read()
            fo.close()
            
            self.extension = os.path.splitext(path)[1][1:].lower()
            self.filename = os.path.basename(path)            
        elif "HTTP_USER_AGENT" not in os.environ or os.environ["HTTP_USER_AGENT"] != "GrabzIt":
            raise Exception("A call originating from a non-GrabzIt server has been detected")

    def __getFile(self):
        if self.file != None:
            return self.file
                
        form = cgi.FieldStorage()

        if "file" in form:
            self.file = form["file"]
                
        return self.file
    
    def toJSON(self):
        if self.getExtension() == 'json':
            return json.loads(self.toString())
        return None
    
    def toXML(self):
        if self.getExtension() == 'xml':
            return ET.ElementTree(ET.fromstring(self.toString()))
        return None    
    
    def toString(self):
        if self.data != None:
            return self.data
            
        obj = self.__getFile()
        if obj != None:
            self.data = obj.file.read()
            
        return self.data
            
    def __str__(self):
        return self.toString()

    def getFilename(self):
        if self.filename != None:
            return self.filename
            
        obj = self.__getFile()
        if obj != None:
            self.filename = obj.filename.split("\\")[-1]
            
        return self.filename

    def getExtension(self):
        if self.extension != None:
            return self.extension
            
        name = self.getFilename()
        if name != None:
            self.extension = name.split(".")[-1]
        
        return self.extension

    def save(self, path):
        try:
            fo = open(path, "wb")
            fo.write(self.toString())
            fo.close()
            return True
        except:
            return False

File: scraper/php/lib/ScrapeResult.php
<?php
namespace GrabzIt\Scraper;

use Exception;

class ScrapeResult
{
    private $extension;
    private $filename;
    private $data;

    public function __construct($path = null)
    {
        if ($path != null)
        {
            $this->data = $this->file_get_contents_utf8($path);
            $this->extension = strtolower(pathinfo($path, PATHINFO_EXTENSION));
            $this->filename = pathinfo($path, PATHINFO_BASENAME);
        }
        else if ($_SERVER['HTTP_USER_AGENT'] != 'GrabzIt')
        {
            throw new \Exception("A call originating from a non-GrabzIt server has been detected");
        }
    }
    
    private function file_get_contents_utf8($fn)
    { 
        $content = file_get_contents($fn); 
        return mb_convert_encoding($content, 'UTF-8', mb_detect_encoding($content, 'UTF-8, ISO-8859-1', true)); 
    }

    public function __toString()
    {
        return $this->toString();
    }

    public function toString()
    {
        if ($this->data == null && $_FILES['file']['error'] == UPLOAD_ERR_OK && is_uploaded_file($_FILES['file']['tmp_name']))
        {
            $this->data = $this->file_get_contents_utf8($_FILES['file']['tmp_name']);
        }
        if ($this->data != null && substr($this->data, 0, 3) == "\xef\xbb\xbf") 
        {
            $this->data = substr($this->data, 3);            
        }
        return $this->data;
    }

    public function toXML()
    {
        if ($this->getExtension() == 'xml')
        {
            return new SimpleXMLElement($this->toString());
        }
        return null;
    }

    public function toJSON()
    {
        if ($this->getExtension() == 'json')
        {
            return json_decode($this->toString());
        }
        return null;
    }

    public function getFilename()
    {
        if ($this->filename == null && $_FILES['file']['name'] != null)
        {
            $this->filename = pathinfo($_FILES['file']['name'], PATHINFO_BASENAME);
        }
        return $this->filename;
    }

    public function getExtension()
    {
        if ($this->extension == null && $_FILES['file']['name'] != null)
        {
            $this->extension = strtolower(pathinfo($_FILES['file']['name'], PATHINFO_EXTENSION));
        }
        return $this->extension;
    }

    public function save($path)
    {
        $dat = $this->toString();
        if ($dat != null)
        {
            return file_put_contents($path, $dat) !== false;
        }
        return false;
    }
}
?>